OpenNefia.Content/UI/UiGroup.cs

```diff
@@ -0,0 +1,212 @@
+using OpenNefia.Core.Input;
+using OpenNefia.Core.IoC;
+using OpenNefia.Core.Rendering;
+using OpenNefia.Core.UI;
+using OpenNefia.Core.UI.Element;
+using OpenNefia.Core.UI.Layer;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenNefia.Core.Maths;
+using OpenNefia.Content.UI.Hud;
+using OpenNefia.Core.Locale;
+using OpenNefia.Core.UserInterface;
+using OpenNefia.Core.Audio;
+
+namespace OpenNefia.Content.UI
+{
+    public class UiGroupArgs<TLayer, TArgs>
+        where TArgs : notnull
+    {
+        public Dictionary<TArgs, TLayer> Layers = new();
+        public TArgs SelectedArgs = default!;
+    }
+
+    public class GroupableUiLayer<TArgs, TResult> : UiLayerWithResult<TArgs, TResult>
+        where TResult : class
+    {
+        public GroupableUiLayer()
+        {
+
+        }
+
+        public virtual void TabEnter()
+        {
+
+        }
+
+        public virtual void TabExit()
+        {
+
+        }
+    }
+
+    public class UiGroup<TLayer, TArgs, TLayerArgs, TResult> : UiLayerWithResult<TArgs, TResult>
+        where TArgs : UiGroupArgs<TLayer, TLayerArgs>
+        where TLayerArgs : notnull
+        where TResult : class
+        where TLayer : GroupableUiLayer<TLayerArgs, TResult>, new()
+    {
+        private Dictionary<TLayerArgs, TLayer> Layers = new();
+        private Dictionary<TLayerArgs, IUiText> Texts = new();
+        private Dictionary<TLayerArgs, (IAssetDrawable Elem, Vector2i Offset)> Icons = new();
+        private UiTopicWindow TabWindow = default!;
+        private IAssetInstance TabDeco;
+
+        private TLayerArgs SelectedArgs = default!;
+        private TLayer SelectedLayer = default!;
+
+        private const int ItemWidth = 50;
+        private const int ExtraTabWindowWidth = 30;
+
+        public UiGroup()
+        {
+            TabWindow = new UiTopicWindow(UiTopicWindow.FrameStyleKind.Three, UiTopicWindow.WindowStyleKind.Four);
+            OnKeyBindDown += OnKeyDown;
+            TabDeco = Assets.Get(Protos.Asset.RadarDeco);
+        }
+
+        public override void Initialize(TArgs args)
+        {
+            base.Initialize(args);
+            SelectedArgs = args.SelectedArgs;
+            Layers = args.Layers;
+            Texts = Layers.ToDictionary(x => x.Key, y => (IUiText)new UiTextOutlined(UiFonts.HUDTabText, GetText(y.Key)));
+            Icons = Layers.ToDictionary(x => x.Key, y => GetIcon(y.Key));
+            ShowSelectedLayer();
+        }
+
+        public override void GrabFocus()
+        {
+            base.GrabFocus();
+            SelectedLayer.GrabFocus();
+        }
+
+        private void OnKeyDown(Core.UI.Element.GUIBoundKeyEventArgs args)
+        {
+            var queryArgs = Layers.Keys.ToList();
+            var index = queryArgs.IndexOf(SelectedArgs);
+            switch(args.Function)
+            {
+                case var next when next == EngineKeyFunctions.UINextTab:
```

Comment:
Since there are only two options, maybe it's better to just do `if (args.Function == EngineKeyFunctions.UINextTab)`?

Reply:
changed in 9eb449f